packages/devtools_app/lib/src/shared/routing.dart

```diff
@@ -215,4 +251,75 @@ class DevToolsRouterDelegate extends RouterDelegate<DevToolsRouteConfiguration>
         {...currentConfiguration!.args, ...?changes},
         {...currentConfiguration!.args},
       );
+
+  /// Checks whether applying [changes] over the current routes state will result
+  /// in any changes.
+  bool _changesState(DevToolsNavigationState? changes) {
+    return currentConfiguration!.state?.hasChanges(changes) ?? false;
+  }
+}
+
+class DevToolsNavigationState {
+  DevToolsNavigationState({
+    required this.kind,
+    required Map<String, String?> state,
+  }) : _state = {
+          '_kind': kind,
+          ...state,
+        };
+
+  factory DevToolsNavigationState.fromJson(Map<String, dynamic> json) =>
+      DevToolsNavigationState._(json.cast<String, String?>());
+
+  DevToolsNavigationState._(this._state) : kind = _state['_kind']!;
+
+  final String kind;
+
+  UnmodifiableMapView<String, String?> get state => UnmodifiableMapView(_state);
+  final Map<String, String?> _state;
+
+  bool hasChanges(DevToolsNavigationState? other) {
+    return !mapEquals(
+      {...state, ...?other?.state},
+      state,
+    );
+  }
+
+  @override
+  String toString() => _state.toString();
+
+  Map<String, dynamic> toJson() => _state;
+}
+
+/// Mixin that gives controllers the ability to respond to changes in router
+/// navigation state.
+mixin RouteStateHandlerMixin on DisposableController {
+  DevToolsRouterDelegate? _delegate;
+
+  @override
+  void dispose() {
+    super.dispose();
+    _delegate?.removeListener(_onRouteStateUpdate);
+  }
+
+  void subscribeToRouterEvents(DevToolsRouterDelegate delegate) {
+    final oldDelegate = _delegate;
+    if (oldDelegate != null) {
+      oldDelegate.removeListener(_onRouteStateUpdate);
+    }
+    delegate.addListener(_onRouteStateUpdate);
+    _delegate = delegate;
+  }
+
+  void _onRouteStateUpdate() {
+    final state = _delegate?.currentConfiguration?.state;
+    if (state == null) return;
+    onRouteStateUpdate(state);
+  }
+
+  /// Perform operations based on changes in navigation state.
+  ///
+  /// This method is only invoked if [subscribeToRouterEvents] has been called on
+  /// this instance with a valid [DevToolsRouterDelegate].
+  void onRouteStateUpdate(DevToolsNavigationState state) {}
```

Comment:
can this be abstract? or another option, can we annotate with mustCallSuper and add an assert in here that `subscribeToRouterEvents` has been invoked for this instance